configure.ac

```diff
@@ -259,26 +259,10 @@ fi
 #
 # libsecp256k1
 #
-# FIXME: This is needed to force libtool to use all object files from secp.
-#        We can only build secp properly by recursively invoking
-#        configure/make, and can't include it as a noinst_ library. Libtool
-#        assumes that such libraries will be installed along with our library
-#        target and so won't force all object files in the library to be
-#        included in ours - despite the fact that we are making a shared
-#        library and linking to a static one. This is broken and we work
-#        around it by hacking the secp objects directly into the library
-#        via the _LDADD variable for wallycore.
-#        We previously achieved this by adding the libsecp256k1.a archive,
-#        but changes to libtool and apples linkers mean that
-#        archives-within-archives no longer work.
-#        Because automake tries to police its users very strictly and fails
-#        hard when flags are passed in this way, we have to substitute the
-#        flags here.
-#        Because libtool both intercepts -Wl and arbitrarily re-orders its
-#        command line inputs, we have to concoct a single expression to
-#        enforce linking that cannot be split, hence the below expression.
-LIBADD_SECP256K1="-Wl,secp256k1/src/libsecp256k1_la-secp256k1.${OBJEXT},secp256k1/src/libsecp256k1_precomputed_la-precomputed_ecmult_gen.${OBJEXT},secp256k1/src/libsecp256k1_precomputed_la-precomputed_ecmult.${OBJEXT}"
-AC_SUBST([LIBADD_SECP256K1])
+libsecp256k1_CFLAGS='-I$(top_srcdir)/src/secp256k1/include'
+libsecp256k1_LIBS='$(top_srcdir)/src/secp256k1/libsecp256k1.la'
+AC_SUBST([libsecp256k1_CFLAGS])
+AC_SUBST([libsecp256k1_LIBS])
```

Comment:
This fix might bear some discussion. The assertion “Libtool […] won't force all object files in the library to be included in ours - despite the fact that we are making a shared library and linking to a static one. This is broken” appears to be incorrect. When I link `libwallycore.la` with libtool in shared mode (i.e., configured with `--enable-shared`) to produce `libwallycore.so`, passing in the `libsecp256k1.la` that was just built in static mode, libtool _does_ copy in all the symbols from the static library and does export the symbols with public visibility in the resulting shared library's dynamic symbol table. In fact I verified using `diff` that the set of `secp256k1_*` dynamic symbols exported from the resulting `libwallycore.so` exactly matches the set of `secp256k1_*` dynamic symbols exported from a `libsecp256k1.so` that results from the same libsecp256k1 sources rebuilt in shared mode, so I don't know how the confusion originated. Maybe there was once a defect in libtool at some time in the past, but if so, it appears to have been fixed.

Reply:
I believe this does not affect all platforms, IIRC Mac and BSD were affected.

Reply:
I don't have any way to test BSD, but I could conceivably test macOS if it's not convenient for you to do it.

Reply:
Okay, so I tested on macOS 10.13, and libtool does the same thing as it does on Linux:

* If configured to build a _static_ libwally-core library, libtool **does not** embed the secp256k1 symbols into `libwallycore.a`, but it **does** list the path of the just-built `libsecp256k1.la` on the `dependency_libs` line in `libwallycore.la`.
* If configured to build a _shared_ libwally-core library, libtool **does** embed the secp256k1 symbols into `libwallycore.1.dylib`, and it **does not** list anything on the `dependency_libs` line in `libwallycore.la`.

So I don't see any problem.

I can't test any BSDs, but I would expect them to have the same behavior as macOS.

Reply:
This workaround was put in place in 2016, so its certainly possible the underlying bug(s) were fixed. I think we have an openbsd box in our internal CI somewhere, I'll need to check the build on that.

Reply:
Another big landmine that we haven't mentioned yet is Windows. Apparently you do support that OS. I have no idea what libtool does with DLLs on such a system. I have no way to test this either, as I haven't used Windows for a couple of decades at this point. 😕

Reply:
autotools on windows isn't worth supporting; too many special cases, bugs and failure modes. For Python we have a build that works by including the amalgamation, downstream app users are expected to statically link given the security and versioning hell that is windows DSO support.

Reply:
I resurrected the `src/amalgamation/combined*.c` sources but relegated them to the `is_windows` pile only.

Reply:
Yeah I'm still pondering the changes to `setup.py`. Note the commit description for the first commit needs changing now that the secp includes have been pulled into their own commit. Also if you move the setting of `libsecp256k1_LIBS` below the existing `AC_SUBST([libsecp256k1_CFLAGS])` this diff will be a single chunk instead of 2.

Reply:
I'm likely to cherry pick these changes and tweak them so actually don't worry about changing this for now.

Reply:
Note, the reason for renaming `LIBADD_SECP256K1` to `libsecp256k1_LIBS` is because the latter is what gets set by `PKG_CHECK_MODULES`. Not relevant in *this* PR, but becomes relevant when `--with-system-secp256k1` is introduced.